### ISSUE TITLE
Configuration stored as connection objects instead of on module singleton

### ISSUE THREAD

**Issue by bdon** [OCHA-DAP/hdx-python-api#7]
Configurations are stored as global state on the `Configuration` module, example:

`Configuration.create(hdx_site='prod')`

I would prefer if Configurations were individually instantiated as a "connection" object. This would make application code easier to unit test and more decoupled. A good example is the AWS Boto API e.g.  
>>> from boto.s3.connection import S3Connection
>>> conn = S3Connection('<aws access key>', '<aws secret key>')

**Comment by mcarans** [comment]
In a previous version of the library, configuration was instantiated as an object and passed around - see: https://github.com/OCHA-DAP/hdx-python-api/commit/22e8f9a6bce529e3be067d19cc74b8eb36c84dd8

I saw that a less experienced user was creating multiple configuration objects in different places in his code and could see it could lead to confusion if he inadvertently created one targeting test and another targeting prod. I thought I would make it easier by not having to pass the configuration around so made the change to a global configuration.

Given that there are less experienced programmers using the library, is there a way to do this that avoids there potentially being multiple configurations active simultaneously?

**Comment by bdon** [comment]
Having multiple active configurations would be desirable in many cases such as injecting a stub object in automated tests. I understand that this would make the code more verbose though, the current arrangement seems fine for most use cases. Closing, thanks.

**Comment by mcarans** [comment]
Reopening

**Comment by mcarans** [comment]
This should be fixed in 1.2.1. Please confirm and close.

**Comment by mcarans** [comment]
@mojodna are you able to look at this?

**Comment by bdon** [comment]
This changeset doesn't seem to affect class methods like Dataset.read_from_hdx which initialize a Dataset via `Dataset()` - correct me if i'm wrong

**Comment by mcarans** [comment]
Good spot, that is a mistake on my part. Will fix.

**Comment by mcarans** [comment]
1.2.2 fixes all the class methods. Please confirm and close.

**Comment by bdon** [comment]
Still having trouble in cases such as this:

```
WARNING:__main__:There is no HDX configuration! Use Configuration.create(**kwargs)
WARNING:__main__:Traceback (most recent call last):
  File "hdx_exports/hdx_export_set.py", line 194, in sync_datasets
    dataset.create_in_hdx()
  File "/venv/lib/python2.7/site-packages/hdx/data/dataset.py", line 477, in create_in_hdx
    if self._dataset_load_from_hdx(self.data['name']):
  File "/venv/lib/python2.7/site-packages/hdx/data/dataset.py", line 348, in _dataset_load_from_hdx
    self._dataset_create_resources_gallery()
  File "/venv/lib/python2.7/site-packages/hdx/data/dataset.py", line 324, in _dataset_create_resources_gallery
    self.old_data['resources'] = self._copy_hdxobjects(self.resources, Resource, 'file_to_upload')
  File "/venv/lib/python2.7/site-packages/hdx/data/hdxobject.py", line 405, in _copy_hdxobjects
    newhdxobject = hdxobjectclass(newhdxobjectdata)
  File "/venv/lib/python2.7/site-packages/hdx/data/resource.py", line 34, in __init__
    super(Resource, self).__init__(initial_data, configuration=configuration)
  File "/venv/lib/python2.7/site-packages/hdx/data/hdxobject.py", line 60, in __init__
    self.configuration = Configuration.read()
  File "/venv/lib/python2.7/site-packages/hdx/configuration.py", line 296, in read
    raise ConfigurationError('There is no HDX configuration! Use Configuration.create(**kwargs)')
ConfigurationError: There is no HDX configuration! Use Configuration.create(**kwargs)
```

It looks like in some cases there is a deep copy of an HDX object, which instantiates a new object via `hdxobjectclass()`, but this does not pass along the original object's configuration, and it fails on `init()`. It looks like this is a bigger rabbit hole than I expected that is adding complexity to the library, I'll explore ways to solve our use case while using the global Configuration object

**Comment by mcarans** [comment]
Reopening for hdxobjectclass

**Comment by mcarans** [comment]
Fixed in 1.2.5. Sorry but you are my only tester for this! Please confirm.

**Comment by mcarans** [comment]
Added unit tests that check passing of configuration from parent dataset to child resources and gallery items
